src/libraries/Microsoft.PowerFx.Core/Public/Values/InMemoryTableValue.cs

```diff
@@ -10,31 +11,54 @@
 namespace Microsoft.PowerFx.Core.Public.Values
 {
     /// <summary>
-    /// In-memory table. 
+    /// In-memory table. Constructed over RecordValues. 
+    /// DValue means items could be error or blank. 
     /// </summary>
-    internal class InMemoryTableValue : TableValue
+    internal class InMemoryTableValue : ObjectCollectionTableValue<DValue<RecordValue>>
     {
-        private readonly IEnumerable<DValue<RecordValue>> _records;
-
-        public override IEnumerable<DValue<RecordValue>> Rows => _records;
+        private readonly RecordType _recordType;
 
         internal InMemoryTableValue(IRContext irContext, IEnumerable<DValue<RecordValue>> records)
-            : base(irContext)
+            : base(irContext, records, null)
         {
             Contract.Assert(IRContext.ResultType is TableType);
             var tableType = (TableType)IRContext.ResultType;
-            var recordType = tableType.ToRecord();
-            _records = records.Select(r =>
+            _recordType = tableType.ToRecord();            
+        }
+
+        protected override DValue<RecordValue> Marshal(DValue<RecordValue> record)
+        {
+            if (record.IsValue)
             {
-                if (r.IsValue)
-                {
-                    return DValue<RecordValue>.Of(new InMemoryRecordValue(IRContext.NotInSource(recordType), r.Value.Fields));
-                }
-                else
-                {
-                    return r;
-                }
-            });
+                return DValue<RecordValue>.Of(
+                    new InMemoryRecordValue(IRContext.NotInSource(_recordType), record.Value.Fields));
```

Comment:
Would be great to get rid of this in the next PR. 
We make table lazy, but still do a full copy for each record.

Reply:
Why do we need to copy each record?